File: Goal.js
class Goal {
    constructor(centerX, centerY, way, length) {
        this.center = new Vector(centerX, centerY);
        this.way = way;
        this.length = length;
        this.depth = 30;

        if (this.way == Way.left) {
            this.depthMargin = this.depth;
        } else if (this.way == Way.right) {
            this.depthMargin = -this.depth;
        }

        this.topPost = new Disc(centerX, centerY - this.length / 2, 7, Infinity, 0.2, 0).setColor(Color.border);
        this.bottomPost = new Disc(centerX, centerY + this.length / 2, 7, Infinity, 0.2, 0).setColor(Color.border);
        this.goalLine = new VerticalBorder(centerX, centerY, this.length, 1);

        this.netBack = new VerticalBorder(centerX - this.depthMargin, centerY, this.length, 0).setColor(Color.black).setCollisionMask([Ball]).extendTo(this.way);
        this.netTop = new HorizontalBorder(centerX - this.depthMargin / 2, centerY - this.length / 2, this.depth, 0).setColor(Color.black).setCollisionMask([Ball]).extendTo(Way.up);
        this.netBottom = new HorizontalBorder(centerX - this.depthMargin / 2, centerY + this.length / 2, this.depth, 0).setColor(Color.black).setCollisionMask([Ball]).extendTo(Way.down);

    }

    checkGoal(ball) {
        if (!(ball instanceof Ball)) {
            return false;
        }
        if (this.way == Way.left) {
            if (ball.center.x < this.center.x && ball.center.y > this.topPost.center.y && ball.center.y < this.bottomPost.center.y) {
                return true;
            } else {
                return false;
            }
        } else if (this.way == Way.right) {
            if (ball.center.x > this.center.x && ball.center.y > this.topPost.center.y && ball.center.y < this.bottomPost.center.y) {
                return true;
            } else {
                return false;
            }
        }
    }
}

File: pyrl/Goal.py
from Vector import Vector
from Way import Way
from Disc import Disc
import math
from Color import Color
from VerticalBorder import VerticalBorder
from HorizontalBorder import HorizontalBorder
from Ball import Ball

class Goal:
    def __init__(self, center_x, center_y, way, length, post_radius):
        self.center = Vector(center_x, center_y)
        self.way = way
        self.length = length
        self.depth = 30
        
        if self.way == Way.left:
            self.depth_margin = self.depth
        elif self.way == Way.right:
            self.depth_margin = -self.depth
            
        self.top_post = Disc(center_x, center_y - self.length / 2, post_radius, math.inf, 0.2, 0).set_color(Color.border)
        self.bottom_post = Disc(center_x, center_y + self.length / 2, post_radius, math.inf, 0.2, 0).set_color(Color.border)
        self.goal_line = VerticalBorder(center_x, center_y, self.length, 1)
        
        self.net_back = VerticalBorder(center_x - self.depth_margin, center_y, self.length, 0).set_color(Color.black).set_collision_mask([Ball]).extend_to(self.way)
        
        self.net_top = HorizontalBorder(center_x - self.depth_margin / 2, center_y - self.length / 2, self.depth, 0).set_color(Color.black).set_collision_mask([Ball]).extend_to(self.way)
        self.net_bottom = HorizontalBorder(center_x - self.depth_margin / 2, center_y + self.length / 2, self.depth, 0).set_color(Color.black).set_collision_mask([Ball]).extend_to(self.way)

        def check_goal(self, ball):
            if not isinstance(ball, Ball):
                return False
            
            if self.way == Way.left:
                
                if ball.center.x < self.center.x and ball.center.y > self.top_post.center.y and ball.center.y < self.bottom_post.center.y:
                    return True
                else:
                    return False
                
            elif self.way == Way.right:
                
                if ball.center.x > self.center.x and ball.center.y > self.top_post.center.y and ball.center.y < self.bottom_post.center.y:
                    return True
                else:
                    return False
                    
